hazelcast/src/main/java/com/hazelcast/map/impl/MapEntries.java

```diff
@@ -35,31 +35,139 @@
  */
 public final class MapEntries implements IdentifiedDataSerializable {
 
+    private static interface ValueWrapper {
```

Comment:
I, like @pveentjer , don't think that we should be using wrappers here; these don't appear to have any performance impact when you're dealing with small numbers of entries, but once you get up to thousands or millions (which is what hazelcast is aiming for), these will present a problem. As it is, you are adding one more reference per entry held.

Iterator overhead is trivial relative to per-entry overhead, particularly when the number of entries is large.

In any case, the goal of your PR is to add an option to skip defensive copies, and I struggle to see how this particular change is relevant.